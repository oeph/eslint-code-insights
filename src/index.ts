import { CodeInsights } from 'bbs-code-insights';
import { Annotation } from 'bbs-code-insights/dist/types';
import { ESLint } from 'eslint';

interface Config {
    /**
     * URL to the Bitbucket Server
     */
    url: string,
    /**
     * Access Token for the Bitbucket Server
     */
    accessToken: string,
    /**
     * Project Key for the Bitbucket repository
     */
    project: string,
    /** 
     * Repository Slug for the Bitbucket repository
     */
    repo: string,
    /**
     * Full Commit Id of the Commit that should be used to add Code Insights
     */
    commitId: string,
    /**
     * Root path of repository on the client.
     * 
     * This is used to create relative paths for Bitbucket
     */
    repositoryRootPath: string,
}

/**
 * Returns the ESLint calculateConfigForFile function.
 * 
 * Used for debugging purposes to check the eslint config. 
 */
function getCalculateConfigForFile(
    filePath: string,
    eslintOptions?: ESLint.Options
): Promise<Record<string, unknown>> {
    const eslint = new ESLint(eslintOptions);
    return eslint.calculateConfigForFile(filePath);
}

/**
 * Executes ESLint and adds the results as Code Insights
 * 
 * @param config config for the code insights
 * @param files files that should be linted
 * @param eslintOptions options for ESLint
 * @returns Promise to the action that creates the code insights
 */
async function addESLintInsights(
    config: Config,
    files: string | string[],
    eslintOptions?: ESLint.Options
): Promise<void> {
    const codeInsights = new CodeInsights({
        url: config.url,
        accessToken: config.accessToken
    }, {
        project: config.project,
        repo: config.repo,
        commitId: config.commitId,
        reportKey: 'oeph.code-insights.eslint',
    });

    const eslint = new ESLint(eslintOptions);
    const results = await eslint.lintFiles(files);
    let errors = 0;
    let fixableErrors = 0;
    let warnings = 0;
    let fixableWarnings = 0;
    const annotations: Array<Annotation> = [];

    results.forEach(result => {
        errors += result.errorCount;
        fixableErrors += result.fixableErrorCount;
        warnings += result.warningCount;
        fixableWarnings += result.fixableWarningCount;

        const path = result.filePath.replace(config.repositoryRootPath, '');

        result.messages.forEach(message => {
            annotations.push({
                path,
                line: message.line,
                message: message.message,
                severity: message.fatal || message.fix === undefined
                    ? 'HIGH'
                    : message.severity === 2
                        ? 'MEDIUM'
                        : 'LOW',
                externalId: `${path}-${message.line}-${message.column}-${message.ruleId}`
            });
        });
    });

    const createdDate = new Date().getTime();
    if (results.length === 0 || errors === 0) {
        return codeInsights.createReport({
            title: 'ESLint Report',
            reporter: 'ESLint Code Insight',
            logoUrl: 'https://eslint.org/img/logo.svg',
            result: 'PASS',
            createdDate,
            data: [{
                title: 'Warning Count',
                type: 'NUMBER',
                value: warnings,
            }, {
                title: 'Fixable Warning Count',
                type: 'NUMBER',
                value: fixableWarnings,
            }]
        });
    } else {
        return codeInsights.createReport({
            title: 'ESLint Report',
            reporter: 'ESLint Code Insight',
            logoUrl: 'https://eslint.org/img/logo.svg',
            result: 'FAIL',
            createdDate,
            data: [{
                title: 'Error Count',
                type: 'NUMBER',
                value: errors,
            }, {
                title: 'Warning Count',
                type: 'NUMBER',
                value: warnings,
            }, {
                title: 'Fixable Error Count',
                type: 'NUMBER',
                value: fixableErrors,
            }, {
                title: 'Fixable Warning Count',
                type: 'NUMBER',
                value: fixableWarnings,
            }]
        }, annotations);
    }
}

export default addESLintInsights;
export { addESLintInsights, getCalculateConfigForFile };