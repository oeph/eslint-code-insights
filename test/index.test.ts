import { getCalculateConfigForFile, addESLintInsights } from '../src/index';

const createReportMock = jest.fn(async () => {
    return Promise.resolve();
});

jest.mock('bbs-code-insights', () => {
    return {
        CodeInsights: jest.fn().mockImplementation(() => {
            return {
                createReport: createReportMock
            };
        })
    };
});

const url = process.env.BBS_HOST || 'https://local-test.example.org';
const accessToken = process.env.BBS_TOKEN || 'super-secret-token';
const repo = process.env.REPO || 'repo';
const project = process.env.PROJECT || 'some-project';
const commitId = process.env.COMMIT_ID || 'e69494baf07d6277393051cabece2ba56bad2a3a';

describe('getCalculateConfigForFile', () => {
    test('standard config', async () => {
        const config = await getCalculateConfigForFile('./index.test.ts');
        expect(config).toMatchSnapshot();
    });
});

describe('addESLintInsights', () => {
    test('should send report', async () => {
        await addESLintInsights({
            url,
            accessToken,
            project,
            repo,
            commitId,
            repositoryRootPath: '/',
        }, 'test/file-to-be-linted.js');

        expect(createReportMock).toBeCalledTimes(1);
    });

    test('should send pass report on clean file', async () => {
        await addESLintInsights({
            url,
            accessToken,
            project,
            repo,
            commitId,
            repositoryRootPath: '/',
        }, 'test/file-to-be-linted.js');

        expect(createReportMock).toBeCalledTimes(1);
        const call: [] = createReportMock.mock.calls[0];
        expect(call).not.toBeUndefined();
        expect(call.length).toBe(1);
        const argument = call.pop();
        expect(argument).toMatchObject({
            title: 'ESLint Report',
            reporter: 'ESLint Code Insight',
            logoUrl: 'https://eslint.org/img/logo.svg',
            result: 'PASS',
            data: [{
                title: 'Warning Count',
                type: 'NUMBER',
                value: 0,
            }, {
                title: 'Fixable Warning Count',
                type: 'NUMBER',
                value: 0,
            }]
        });
    });

    test('should send fail report on file with errors', async () => {
        await addESLintInsights({
            url,
            accessToken,
            project,
            repo,
            commitId,
            repositoryRootPath: '/',
        }, 'test/file-to-be-linted-w-errors.js');

        expect(createReportMock).toBeCalledTimes(1);
        const call: [] = createReportMock.mock.calls[0];
        expect(call).not.toBeUndefined();
        expect(call.length).toBe(2);
        const annotationArgument = call.pop();
        const reportArgument = call.pop();
        expect(reportArgument).toMatchObject({
            title: 'ESLint Report',
            reporter: 'ESLint Code Insight',
            logoUrl: 'https://eslint.org/img/logo.svg',
            result: 'FAIL',
            data: [{
                title: 'Error Count',
                type: 'NUMBER',
                value: 2,
            }, {
                title: 'Warning Count',
                type: 'NUMBER',
                value: 2,
            }, {
                title: 'Fixable Error Count',
                type: 'NUMBER',
                value: 1,
            }, {
                title: 'Fixable Warning Count',
                type: 'NUMBER',
                value: 0,
            }]
        });

        expect(annotationArgument).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    'line': 1,
                    'message': '\'someVar\' is assigned a value but never used.',
                    'severity': 'HIGH'
                })
            ])
        );

        expect(annotationArgument).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    'line': 3,
                    'message': '\'aFunction\' is defined but never used.',
                    'severity': 'HIGH'
                })
            ])
        );

        expect(annotationArgument).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    'line': 4,
                    'message': '\'console\' is not defined.',
                    'severity': 'HIGH'
                })
            ])
        );

        expect(annotationArgument).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    'line': 4,
                    'message': 'Missing semicolon.',
                    'severity': 'MEDIUM'
                })
            ])
        );
    });
});