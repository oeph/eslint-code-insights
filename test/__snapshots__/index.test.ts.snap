// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCalculateConfigForFile standard config 1`] = `
Object {
  "env": Object {
    "es6": true,
    "jest": true,
  },
  "globals": Object {},
  "ignorePatterns": Array [],
  "noInlineConfig": undefined,
  "parser": "C:\\\\Users\\\\Philipp\\\\Projekte\\\\github\\\\eslint-code-insight\\\\node_modules\\\\@typescript-eslint\\\\parser\\\\dist\\\\index.js",
  "parserOptions": Object {
    "sourceType": "module",
  },
  "plugins": Array [
    "@typescript-eslint",
  ],
  "reportUnusedDisableDirectives": undefined,
  "rules": Object {
    "@typescript-eslint/adjacent-overload-signatures": Array [
      "error",
    ],
    "@typescript-eslint/ban-ts-comment": Array [
      "error",
    ],
    "@typescript-eslint/ban-types": Array [
      "error",
    ],
    "@typescript-eslint/explicit-module-boundary-types": Array [
      "warn",
    ],
    "@typescript-eslint/no-array-constructor": Array [
      "error",
    ],
    "@typescript-eslint/no-empty-function": Array [
      "error",
    ],
    "@typescript-eslint/no-empty-interface": Array [
      "error",
    ],
    "@typescript-eslint/no-explicit-any": Array [
      "warn",
    ],
    "@typescript-eslint/no-extra-non-null-assertion": Array [
      "error",
    ],
    "@typescript-eslint/no-extra-semi": Array [
      "error",
    ],
    "@typescript-eslint/no-inferrable-types": Array [
      "error",
    ],
    "@typescript-eslint/no-misused-new": Array [
      "error",
    ],
    "@typescript-eslint/no-namespace": Array [
      "error",
    ],
    "@typescript-eslint/no-non-null-asserted-optional-chain": Array [
      "error",
    ],
    "@typescript-eslint/no-non-null-assertion": Array [
      "warn",
    ],
    "@typescript-eslint/no-this-alias": Array [
      "error",
    ],
    "@typescript-eslint/no-unused-vars": Array [
      "warn",
    ],
    "@typescript-eslint/no-var-requires": Array [
      "error",
    ],
    "@typescript-eslint/prefer-as-const": Array [
      "error",
    ],
    "@typescript-eslint/prefer-namespace-keyword": Array [
      "error",
    ],
    "@typescript-eslint/triple-slash-reference": Array [
      "error",
    ],
    "constructor-super": Array [
      "off",
    ],
    "for-direction": Array [
      "error",
    ],
    "getter-return": Array [
      "off",
    ],
    "linebreak-style": Array [
      "error",
      "unix",
    ],
    "no-array-constructor": Array [
      "off",
    ],
    "no-async-promise-executor": Array [
      "error",
    ],
    "no-case-declarations": Array [
      "error",
    ],
    "no-class-assign": Array [
      "error",
    ],
    "no-compare-neg-zero": Array [
      "error",
    ],
    "no-cond-assign": Array [
      "error",
    ],
    "no-const-assign": Array [
      "off",
    ],
    "no-constant-condition": Array [
      "error",
    ],
    "no-control-regex": Array [
      "error",
    ],
    "no-debugger": Array [
      "error",
    ],
    "no-delete-var": Array [
      "error",
    ],
    "no-dupe-args": Array [
      "off",
    ],
    "no-dupe-class-members": Array [
      "off",
    ],
    "no-dupe-else-if": Array [
      "error",
    ],
    "no-dupe-keys": Array [
      "off",
    ],
    "no-duplicate-case": Array [
      "error",
    ],
    "no-empty": Array [
      "error",
    ],
    "no-empty-character-class": Array [
      "error",
    ],
    "no-empty-function": Array [
      "off",
    ],
    "no-empty-pattern": Array [
      "error",
    ],
    "no-ex-assign": Array [
      "error",
    ],
    "no-extra-boolean-cast": Array [
      "error",
    ],
    "no-extra-semi": Array [
      "off",
    ],
    "no-fallthrough": Array [
      "error",
    ],
    "no-func-assign": Array [
      "off",
    ],
    "no-global-assign": Array [
      "error",
    ],
    "no-import-assign": Array [
      "off",
    ],
    "no-inner-declarations": Array [
      "error",
    ],
    "no-invalid-regexp": Array [
      "error",
    ],
    "no-irregular-whitespace": Array [
      "error",
    ],
    "no-misleading-character-class": Array [
      "error",
    ],
    "no-mixed-spaces-and-tabs": Array [
      "error",
    ],
    "no-new-symbol": Array [
      "off",
    ],
    "no-obj-calls": Array [
      "off",
    ],
    "no-octal": Array [
      "error",
    ],
    "no-prototype-builtins": Array [
      "error",
    ],
    "no-redeclare": Array [
      "off",
    ],
    "no-regex-spaces": Array [
      "error",
    ],
    "no-self-assign": Array [
      "error",
    ],
    "no-setter-return": Array [
      "off",
    ],
    "no-shadow-restricted-names": Array [
      "error",
    ],
    "no-sparse-arrays": Array [
      "error",
    ],
    "no-this-before-super": Array [
      "off",
    ],
    "no-undef": Array [
      "off",
    ],
    "no-unexpected-multiline": Array [
      "error",
    ],
    "no-unreachable": Array [
      "off",
    ],
    "no-unsafe-finally": Array [
      "error",
    ],
    "no-unsafe-negation": Array [
      "off",
    ],
    "no-unused-labels": Array [
      "error",
    ],
    "no-unused-vars": Array [
      "off",
    ],
    "no-useless-catch": Array [
      "error",
    ],
    "no-useless-escape": Array [
      "error",
    ],
    "no-var": Array [
      "error",
    ],
    "no-with": Array [
      "error",
    ],
    "prefer-const": Array [
      "error",
    ],
    "prefer-rest-params": Array [
      "error",
    ],
    "prefer-spread": Array [
      "error",
    ],
    "quotes": Array [
      "error",
      "single",
    ],
    "require-yield": Array [
      "error",
    ],
    "semi": Array [
      "error",
      "always",
    ],
    "use-isnan": Array [
      "error",
    ],
    "valid-typeof": Array [
      "off",
    ],
  },
  "settings": Object {},
}
`;
